src/array.cr

```diff
@@ -1545,22 +1544,55 @@ class Array(T)
     self
   end
 
+  # Returns `self` with all the elements shifted `n` times.
+  #
+  # ```
+  # a = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+  # a.rotate! # => [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]
+  # a.rotate!(1) # => [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]
+  # a.rotate!(3) # => [3, 4, 5, 6, 7, 8, 9, 0, 1, 2]
+  # ```
   def rotate!(n = 1)
     return self if size == 0
+    # This is broken into cases for perfromance (#8515)
     n %= size
-    return self if n == 0
-    if n <= size // 2
+    if n == 0
+    elsif n == 1
+      tmp = self[0]
+      @buffer.move_from(@buffer + n, size - n)
+      self[-1] = tmp
+    elsif n == -1
+      tmp = self[-1]
+      @buffer.move_from(@buffer + n, size - n)
+      self[0] = tmp
+    elsif n <= SMALL_ARRAY_SIZE
+      tmp = StaticArray(T, SMALL_ARRAY_SIZE).new{|i| self[i % size] }
```

Comment:
Maybe we should allocate `SMALL_ARRAY_SIZE` but not initialize it with every element in the range? Also `self[i % size]` will do a lot of bound checks, plus `%` is a bit slower than `remainder`. Not necessary to change it now, maybe an optimization for later.

Reply:
I did that because StaticArray size cannot be declared at runtime.
```crystal
StaticArray(Int32, n).new(0)
```
```text
expecting token ')', not 'n'
```

Reply:
You can do:

```crystal
tmp = uninitialized Int32[SMALL_ARRAY_SIZE]
# now only copy n elements, not SMALL_ARRAY_SIZE, into it
```